src/Confluent.Kafka/Diagnostics.cs

```diff
@@ -0,0 +1,75 @@
+﻿// Copyright 2022 Confluent Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Refer to LICENSE for more information.
+
+using System.Diagnostics;
+using System.Text;
+
+namespace Confluent.Kafka
+{
+    /// <summary>
+    /// Implements Activity objects with OpenTelemetry messaging tags for instrumentation
+    /// </summary>
+    internal static class Diagnostics
+    {
+        private const string ActivitySourceName = "Confluent.Kafka";
+        public static ActivitySource ActivitySource { get; } = new ActivitySource(ActivitySourceName);
+
+        /// <summary>
+        /// Provides an Activity object for the Producer with OpenTelemetry messaging tags for instrumentation
+        /// </summary>
+        internal static class Producer
+        {
+            private const string ActivityName = ActivitySourceName + ".MessageProduced";
+
+            internal static Activity Start<TKey, TValue>(TopicPartition topicPartition, Message<TKey, TValue> message)
+            {
+                Activity activity = ActivitySource.StartActivity(ActivityName);
+
+                if (activity == null)
+                    return null;
+
+                using (activity)
+                {
+                    activity?.AddDefaultOpenTelemetryTags(topicPartition, message);

```

Comment:
Don't need the null-conditional here (`activity?`) because line 41 already verified it is non-null.